pkg/endpoint/endpoint.go

```diff
@@ -1804,11 +1805,68 @@ func (e *Endpoint) replaceIdentityLabels(l pkgLabels.Labels) int {
 	return rev
 }
 
-// LeaveLocked removes the endpoint's directory from the system. Must be called
-// with Endpoint's mutex AND BuildMutex locked.
-func (e *Endpoint) LeaveLocked(owner Owner, proxyWaitGroup *completion.WaitGroup) []error {
+// Delete deletes the endpoint and all associated resources.
+func (e *Endpoint) Delete(owner Owner, releaseIP bool) []error {
 	errors := []error{}
 
+	e.UnconditionalLock()
```

Comment:
Specifically in the case of delete, shouldn't BuildMutex also be held? This way we would wait for any in-process build to complete before we delete files, but we also stop any builds from starting (since once we unlock BuildMutex they would then check the state via a LockAlive and stop building?

I'm probably quite incorrect, however, since we unlock BuildMutex just before calling Delete in the one place we call it (in the old code, however, we held it).

Reply:
I've moved this code out into a separate PR as it needs further work. Holding the BuildLock is not required as no build can start without before holding the endpoint lock and without verifying that the endpoint state is not in disconnecting. So it's fine to only hold the build lock while we transition the state to disconnecting.